pkg/validator/promql_expression_helpers.go

```diff
@@ -34,6 +35,63 @@ func getExpressionUsedLabels(expr string) ([]string, error) {
 	return slices.Compact(usedLabels), nil
 }
 
+// Returns true in case metric name selector matches given regexp and a list of other labels beside __name__, false and empty list otherwise.
+func labelsUsedInSelectorForMetricRegexp(selector *parser.VectorSelector, metricRegexp string) (usedLabels []string, metricUsed bool) {
+	for _, m := range selector.LabelMatchers {
+		if m.Name == "__name__" && m.Type == labels.MatchEqual && regexp.MustCompile("^"+metricRegexp+"$").MatchString(m.Value) {
+			metricUsed = true
+			continue
+		}
+		usedLabels = append(usedLabels, m.Name)
+	}
+	return usedLabels, metricUsed
+}
+
+// Returns a list of labels which are used in given expr in relation to given metric.
+// Beside labels within vector selector itself, it adds labels used in Aggregate expressions and labels used in Binary expression.
+// For Binary expressions it may report false positives as the current implementation does not consider on which side of group_left/group_right is the given metric.
+func getExpressionUsedLabelsForMetricRegexp(expr, metricRegexp string) ([]string, error) {
```

Comment:
This whole function looks like a lot of logic duplication from the `getExpressionUsedLabels` above

Maybe adding a new `getExpressionUsedLabelsForMetric(expr, regexp)` and making `getExpressionUsedLabels` just wrapper for `getExpressionUsedLabelsForMetric(expr, ".*")` could suffice?

Reply:
Those two functions seemed to me different, as their approach to going through AST is not the same, but I guess you are right, and really `getExpressionUsedLabels` can be seen as a special case of the other one.

Refactored as suggested.